RHDEVDOCS-4426 - Loki Retention (#55237)
openshift/openshift-docs: modules/logging-loki-retention.adoc

// Module included in the following assemblies:
//
:_content-type: PROCEDURE
[id="logging-loki-retention_{context}"]
= Enabling stream-based retention with Loki
With Logging version 5.6 and higher, you can configure retention policies based on log streams. Rules for these may be set globally, per tenant, or both. If you configure both, tenant rules apply before global rules.

. To enable stream-based retention,  create or edit the `LokiStack` custom resource (CR):

include::snippets/logging-create-apply-cr-snip.adoc[lines=9..12]

. You can refer to the examples below to configure your LokiStack CR.

.Example global stream-based retention
[source,yaml]
----
apiVersion: loki.grafana.com/v1
kind: LokiStack
metadata:
  name: lokistack-sample
  namespace: openshift-logging
spec:

>     limits:

Reviewer: Too many blanks before `limits`, here should be:
```
spec:
  limits:
    global:
```

// Module included in the following assemblies:
//
:_content-type: REFERENCE
[id="logging-loki-retention_{context}"]

.Example LokiStack CR 

[source,yaml]
----

> apiVersion: loki.grafana.com/v1

Reviewer 2: This look like the correct LokiStack for 'per-tenant stream based retention'. We can also define global stream based retention' for LokiStack.

For e.g.
 ```
apiVersion: loki.grafana.com/v1
 kind: LokiStack
 metadata:
   name: lokistack-sample
   namespace: openshift-logging
 spec:
   limits:
     global:
       retention:
         days: 2
         streams:
         - days: 4
           priority: 1
           selector: '{kubernetes_namespace_name=~"test.+"}'
         - days: 1
           priority: 1
           selector: '{log_type="application"}'
         - days: 15
           priority: 1
           selector: '{log_type="audit"}'
   managementState: Managed
   replicationFactor: 1
   size: 1x.small
   storage:
    schemas:
    - effectiveDate: "2020-10-11"
       version: v11
    secret:
      name: gcs-secret
      type: gcs
  storageClassName: standard
  tenants:
    mode: openshift-logging
```

// Module included in the following assemblies:
//
:_content-type: PROCEDURE
[id="logging-loki-retention_{context}"]

> = Enabling stream based retention with Loki

Reviewer 3: ```suggestion
= Enabling stream-based retention with Loki
```

> With Logging version 5.6 and higher, you can configure retention policies based on stream globally, per tenant, or both. If you configure both, tenant rules apply before global rules. 

Reviewer 3: "based on stream globally" sounds a bit odd.

> With Logging version 5.6 and higher, you can configure retention policies based on stream. Rules for these may be set globally, per tenant, or both. If you configure both, tenant rules apply before global rules.

Reviewer 4: ```suggestion
With Logging version 5.6 and higher, you can configure retention policies based on log streams. Rules for these may be set globally, per tenant, or both. If you configure both, tenant rules apply before global rules.
```

> . To enable stream based retention,  create or edit the `LokiStack` custom resource:

Reviewer 3: ```suggestion
. To enable stream-based retention, create or edit the `LokiStack` custom resource (CR):
```

  kind: LokiStack
  metadata:
    name: lokistack-dev
  spec:
    size: 1x.small
    storage:
      secret:
        name: test
        type: s3
    storageClassName: gp2
    retention: <1>
      deleteDelay: <2>
    limits:
      global: <3>
        retentionLimits:
          periodDays: 31
          stream:
            - selector:
                name: namespace
                type: equal
                value: dev
              priority: 1
              periodDays: 1
      tenants: <4>
        tenanta:
          retentionLimits:
            periodDays: 7
            stream:
              - selector:
                  name: namespace
                  type: equal <5>
                  value: prod
                priority: 2
                periodDays: 14
              - selector:
                  name: container
                  type: equal
                  value: loki
                priority: 1
                periodDays: 3
        tenantb:
          retentionLimits:
            periodDays:
            stream:
              - selector:
                  name: container
                  type: equal
                  value: nginx
                priority: 1

>                 periodDays: 1

Reviewer: This is not a valid LokiStack CR, many fields are not in the lokistack CRD: https://github.com/openshift/loki/blob/release-5.6/operator/bundle/manifests/loki.grafana.com_lokistacks.yaml.

Some examples:
1. Per-Tenant stream-based retention
```
apiVersion: loki.grafana.com/v1
kind: LokiStack
metadata:
  name: lokistack-sample
  namespace: openshift-logging
spec:
  limits:
    global:
      retention:
        days: 20
    tenants:
      application:
        retention:
          days: 1
          streams:
            - days: 4
              selector: '{kubernetes_namespace_name=~"test.+"}'
      infrastructure:
        retention:
          days: 5
          streams:
            - days: 1
              selector: '{kubernetes_namespace_name=~"openshift-cluster.+"}'
  managementState: Managed
  replicationFactor: 1
  size: 1x.extra-small
  storage:
    schemas:
    - effectiveDate: "2020-10-11"
      version: v11
    secret:
      name: gcs-secret
      type: gcs
  storageClassName: standard
  tenants:
    mode: openshift-logging 
```
2. Global stream-based retention
```
 apiVersion: loki.grafana.com/v1
 kind: LokiStack
 metadata:
   name: lokistack-sample
   namespace: openshift-logging
 spec:
   limits:
     global:
       retention:
         days: 20
         streams:
         - days: 4
           priority: 1
           selector: '{kubernetes_namespace_name=~"test.+"}'
         - days: 1
           priority: 1
           selector: '{log_type="infrastructure"}'
   managementState: Managed
```

----
<1> Retention is enabled in the cluster when this block is added to the CR.
<2> Time after which the compactor will delete marked chunks.

> <3> Sets retention policy for all logs.

Reviewer 4: ```suggestion
<3> Sets retention policy for all log streams.
```

We should highlight in a note below that this field is not to be used as the global retention policy of 30days that OpenShift Logging includes as a maximum supported policy. The global 30 days policy across all stored logs has to be defined on the Object Storage provider side for the LokiStack bucket. The key wording here is:
- OpenShift Logging 30 days retention -> Affects **stored logs** on object storage -> Configuration on the object storage provider only
- The field mentioned in `<3>` is a global **log stream** retention policy and applying it is at best effort by the compactor. In addition if someone mixes tenant with global retention policies the global policy is overridden anyway.
